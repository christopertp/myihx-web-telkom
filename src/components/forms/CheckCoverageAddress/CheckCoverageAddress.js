import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { Field } from 'redux-form';
import Button from '../../elements/Button';
import Spinner from '../../elements/Spinner';
import Autocomplete from '../../fields/Autocomplete';
import TextField from '../../fields/Text';
import styles from './styles.scoped.css';

export default function CheckCoverageAddress(props) {
  const { address, change, handleSubmit, invalid } = props;
  const dispatch = useDispatch();
  const data = useSelector(s => s.checkCoverage);
  const formData = useSelector(s => s.form.checkCoverageAddress.values);
  const { cities, districts, fetchAddress, isLoading, message, provinces, streets } = data;
  const { cityCode, districtCode, provinceCode } = formData;

  const [mapCities, resetCities] = useMappedAddress(cities);
  const [mapDistricts, resetDistricts] = useMappedAddress(districts);
  const [mapProvinces] = useMappedAddress(provinces);
  const [mapStreets, resetStreets] = useMappedAddress(streets);

  useEffect(() => {
    dispatch(fetchAddress('provinces'));
    change('addressDescription', address);
  }, []);

  const onSearch = (name, value) => {
    const fn = {
      provinces: () => fetchAddress('provinces', value),
      cities: () => fetchAddress('cities', provinceCode, value),
      districts: () => fetchAddress('districts', cityCode, value),
      streets: () => fetchAddress('streets', cityCode, districtCode, value),
    };
    dispatch(fn[name]());
  };

  useEffect(() => {
    provinceCode && dispatch(fetchAddress('cities', provinceCode));
    resetCities();
    resetDistricts();
    resetStreets();
    change('cityCode', '');
    change('districtCode', '');
    change('streetCode', '');
  }, [provinceCode]);

  useEffect(() => {
    cityCode && dispatch(fetchAddress('districts', cityCode));
    resetDistricts();
    resetStreets();
    change('districtCode', '');
    change('streetCode', '');
  }, [cityCode]);

  useEffect(() => {
    if (districtCode) {
      dispatch(fetchAddress('streets', cityCode, districtCode));
      change('postalCode', districts.find(i => i.id === districtCode).postalCode);
    }
    resetStreets();
    change('streetCode', '');
  }, [districtCode]);

  const loadingPlaceholder = (key, placeholder) => isLoading[key] ? 'Loading...' : placeholder;
  const selectDisabled = prevCode => !prevCode;

  const inputProps = [
    {
      label: 'PROVINCE',
      placeholder: loadingPlaceholder('p', 'Choose your province'),
    },
    {
      disabled: selectDisabled(provinceCode),
      label: 'CITY',
      placeholder: loadingPlaceholder('c', 'Choose'),
    },
    {
      disabled: selectDisabled(cityCode),
      label: 'DISTRICT',
      placeholder: loadingPlaceholder('d', 'Choose'),
    },
    {
      disabled: selectDisabled(districtCode),
      label: 'NEAREST ADDRESS',
      placeholder: loadingPlaceholder('s', 'Choose nearest address'),
    },
    {
      label: 'DETAILED ADDRESS',
      placeholder: 'Enter your address',
    },
    {
      label: 'RT/RW (OPTIONAL)',
      placeholder: 'Enter number',
    },
    {
      disabled: true,
      label: 'ZIP CODE',
      placeholder: 'Enter number',
    },
  ];

  return (
    <form className={styles.root} onSubmit={handleSubmit}>
      <Field component={Autocomplete} inputProps={inputProps[0]} name="provinceCode" onSearch={v => onSearch('provinces', v)} options={mapProvinces} />
      <section className={styles.half}>
        <div><Field component={Autocomplete} inputProps={inputProps[1]} name="cityCode" onSearch={v => onSearch('cities', v)} options={mapCities} /></div>
        <div><Field component={Autocomplete} inputProps={inputProps[2]} name="districtCode" onSearch={v => onSearch('districts', v)} options={mapDistricts} /></div>
      </section>
      <Field component={Autocomplete} inputProps={inputProps[3]} name="streetCode" onSearch={v => onSearch('streets', v)} options={mapStreets} />
      <Field component={TextField} inputProps={inputProps[4]} name="addressDescription" />
      <section className={styles.half}>
        <div><Field component={TextField} inputProps={inputProps[5]} name="rtRw" normalize={normalizeRtRw} /></div>
        <div><Field component={TextField} inputProps={inputProps[6]} name="postalCode" /></div>
      </section>
      {(message && message !== 'PT2') && <small className={styles.error}>{message}</small>}
      <Button disabled={invalid || isLoading.submit} type="submit">
        {isLoading.submit ? <Spinner /> : 'Next'}
      </Button>
    </form>
  );
}

CheckCoverageAddress.defaultProps = {
  address: '',
  change: () => { },
  handleSubmit: () => { },
  invalid: false,
};

CheckCoverageAddress.propTypes = {
  address: PropTypes.string,
  change: PropTypes.func,
  handleSubmit: PropTypes.func,
  invalid: PropTypes.bool,
};

export function normalizeRtRw(v, pv) {
  const [rt, rw] = v.split('/');
  if (v.match(/^\d\d\d\d$/)) {
    return `${v.slice(0, 3)}/${v.slice(-1)}`;
  } else if (rt && rt.length > 3) {
    return pv;
  } else if (rw && rw.length > 3) {
    return pv;
  } else if (v.match(/[^0-9/]/)) {
    return pv;
  } else {
    return v;
  }
}

export function useMappedAddress(data) {
  const [mapData, setMapData] = useState([]);
  useEffect(() => {
    setMapData(data.map(i => ({ text: i.name, value: `${i.id}` })));
  }, [data]);
  return [mapData, () => setMapData([])];
}
