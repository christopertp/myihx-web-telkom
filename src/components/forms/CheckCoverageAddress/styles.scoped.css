.root > button {
  margin-top: 2rem;
  width: 100%;
}

.root > label:not(:first-of-type), .half {
  margin-top: 1.5rem;
}

.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 1rem;
}

.half {
  display: flex;
  justify-content: space-between;
}

.half > div { flex: 1; }
.half > div:first-child { margin-right: 0.75rem; }

@media only screen and (max-width: 48em) {
  .root {
    width: 100%;
  }

  .root > label:not(:first-of-type), .half {
    margin-top: 1rem;
  }
}
