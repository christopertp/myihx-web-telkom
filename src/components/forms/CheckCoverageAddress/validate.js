export default function validate(values) {
  const {
    provinceCode,
    cityCode,
    districtCode,
    streetCode,
    addressDescription,
    postalCode,
  } = values;
  return {
    provinceCode: !provinceCode ? 'Please select province' : '',
    cityCode: !cityCode ? 'Please select city' : '',
    districtCode: !districtCode ? 'Please select district' : '',
    streetCode: !streetCode ? 'Please select nearest address' : '',
    addressDescription: !addressDescription ? 'Please enter the address' : '',
    postalCode: !postalCode ? 'Please enter zip code' : '',
  };
}
