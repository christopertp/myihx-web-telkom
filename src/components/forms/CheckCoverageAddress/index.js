import { reduxForm } from 'redux-form';
import CheckCoverageAddress from './CheckCoverageAddress' ;
import validate from './validate';

export default reduxForm({
  initialValues: {
    provinceCode: '',
    cityCode: '',
    districtCode: '',
    streetCode: '',
    addressDescription: '',
    rtRw: '',
    postalCode: '',
  },
  form: 'checkCoverageAddress',
  validate,
})(CheckCoverageAddress);
