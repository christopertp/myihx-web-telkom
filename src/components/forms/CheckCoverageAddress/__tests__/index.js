import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../elements/Spinner';
import CheckCoverageAddress, { normalizeRtRw, useMappedAddress } from '../CheckCoverageAddress';

const defaultState = {
  cities: [],
  districts: [],
  fetchAddress: () => { },
  isLoading: false,
  message: '',
  provinces: [],
  resetMessage: () => { },
  streets: [],
};

useSelector.mockImplementation(fn => {
  fn({ form: { checkCoverageAddress: '' } });
  return defaultState;
});

describe('src/components/forms/CheckCoverageAddress', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<CheckCoverageAddress />);
    CheckCoverageAddress.defaultProps.change();
    CheckCoverageAddress.defaultProps.handleSubmit();
    expect(tree).toMatchSnapshot();
  });

  test('loading & error message', () => {
    useSelector.mockImplementationOnce(() => ({
      ...defaultState,
      message: 'error',
      isLoading: { p: true, submit: true },
    }));
    const result = CheckCoverageAddress(CheckCoverageAddress.defaultProps);
    const formChildren = result.props.children;
    expect(formChildren[5].props.children).toBe('error');
    expect(formChildren[6].props.children.type).toBe(Spinner);
  });

  test('onSearch & useEffect with code', () => {
    useSelector.mockImplementationOnce(() => ({
      ...defaultState,
      districts: [{ id: '1' }],
    }))
      .mockImplementationOnce(() => ({
        cityCode: '1',
        districtCode: '1',
        provinceCode: '1',
      }));
    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);
    const result = CheckCoverageAddress(CheckCoverageAddress.defaultProps);
    const formChildren = result.props.children;
    formChildren[0].props.onSearch();
    formChildren[1].props.children[0].props.children.props.onSearch();
    formChildren[1].props.children[1].props.children.props.onSearch();
    formChildren[2].props.onSearch();
    expect(dispatch).toHaveBeenCalledTimes(8);
  });

  test('normalizeRtRw', () => {
    const result1 = normalizeRtRw('0010', '0');
    expect(result1).toBe('001/0');
    const result2 = normalizeRtRw('0010/0', '0');
    expect(result2).toBe('0');
    const result3 = normalizeRtRw('001/0000', '0');
    expect(result3).toBe('0');
    const result4 = normalizeRtRw('a', '0');
    expect(result4).toBe('0');
    const result5 = normalizeRtRw('001/002', '001/00');
    expect(result5).toBe('001/002');
  });

  test('useMappedAddress', () => {
    const setMapData = jest.fn();
    useState.mockImplementationOnce(v => [v, setMapData]);
    const result = useMappedAddress([{ name: 'name', id: 1 }]);
    result[1]();
    expect(result[0].length).toBe(0);
    expect(setMapData).toHaveBeenCalledWith([]);
  });
});
