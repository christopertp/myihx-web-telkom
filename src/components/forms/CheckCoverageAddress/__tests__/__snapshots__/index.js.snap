// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/forms/CheckCoverageAddress render 1`] = `
<form
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "PROVINCE",
        "placeholder": "Choose your province",
      }
    }
    name="provinceCode"
    onSearch={[Function]}
    options={Array []}
  />
  <section>
    <div>
      <Field
        component={[Function]}
        inputProps={
          Object {
            "disabled": true,
            "label": "CITY",
            "placeholder": "Choose",
          }
        }
        name="cityCode"
        onSearch={[Function]}
        options={Array []}
      />
    </div>
    <div>
      <Field
        component={[Function]}
        inputProps={
          Object {
            "disabled": true,
            "label": "DISTRICT",
            "placeholder": "Choose",
          }
        }
        name="districtCode"
        onSearch={[Function]}
        options={Array []}
      />
    </div>
  </section>
  <Field
    component={[Function]}
    inputProps={
      Object {
        "disabled": true,
        "label": "NEAREST ADDRESS",
        "placeholder": "Choose nearest address",
      }
    }
    name="streetCode"
    onSearch={[Function]}
    options={Array []}
  />
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "DETAILED ADDRESS",
        "placeholder": "Enter your address",
      }
    }
    name="addressDescription"
  />
  <section>
    <div>
      <Field
        component={[Function]}
        inputProps={
          Object {
            "label": "RT/RW (OPTIONAL)",
            "placeholder": "Enter number",
          }
        }
        name="rtRw"
        normalize={[Function]}
      />
    </div>
    <div>
      <Field
        component={[Function]}
        inputProps={
          Object {
            "disabled": true,
            "label": "ZIP CODE",
            "placeholder": "Enter number",
          }
        }
        name="postalCode"
      />
    </div>
  </section>
  
  <Button
    className=""
    disabled={false}
    onClick={null}
    to=""
    type="submit"
    variant="red"
  >
    Next
  </Button>
</form>
`;
